File: PAT (Advanced Level) Practise/1090_Highest Price in Supply Chain (25).py
# Pyton 超时
n, p, r = map(float, raw_input().split())
relation = map(int, raw_input().split())
adjList = {}
root = -1
for idx, val in enumerate(relation):
    adjList.setdefault(val, []).append(idx)
    if val == -1:
        root = idx

max_dep = 0
max_dep_cnt = 0

q = []
q.append((root, 0))
while len(q) != 0:
    cur, dep = q.pop(0)
    if dep > max_dep:
        max_dep, max_dep_cnt = dep, 1
    elif dep == max_dep:
        max_dep_cnt += 1
    if adjList.get(cur):
        for v in adjList.get(cur):
            q.append((v, dep+1))

print "%.2f %d" % (p*(1+r/100)**max_dep, max_dep_cnt)

File: PAT (Advanced Level) Practise/1090_Highest Price in Supply Chain (25).cpp
#include <bits/stdc++.h>
using namespace std;
const int maxn = 100005;
vector<int> adjList[maxn];
int main() {
    int n, v, root;
    double p, r;
    scanf("%d %lf %lf", &n, &p, &r);
    // 预处理出，边邻接表，保存root
    for (int i = 0; i < n; i++) {
        scanf("%d", &v);
        adjList[v].push_back(i);
        if (v == -1) root = i;
    }
    // 从root开始BFS，并标记每层深度，i记录最深的节点和数量
    int max_dep = 0, max_dep_cnt = 0;
    queue<pair<int, int>> q;
    q.push(make_pair(root, 0));
    while (!q.empty()) {
        pair<int, int> p = q.front();
        int cur = p.first, dep = p.second;
        q.pop();
        if (dep > max_dep) {
            max_dep = dep;
            max_dep_cnt = 1;
        }
        else if (dep == max_dep) {
            max_dep_cnt += 1;
        }
        for (int i = 0; i < adjList[cur].size(); i++)
            q.push(make_pair(adjList[cur][i], dep + 1));
    }
    printf("%.2lf %d\n", p*pow((1 + r / 100), max_dep), max_dep_cnt);
    return 0;
}